### PR TITLE
32 dialog strict type

### PR DESCRIPTION
**Pull request by wwalmnes**
Fixes https://github.com/nrkno/core-components/issues/31

I'm not completely satisfied with `strict={'true'| /*or omit attribute*/} ` on line 172 in `core-dialog.md`. I want to either send in `strict={true|false}` (as booleans and not strings), but I'm getting an error. Any suggestions @eirikbacker

### CODE REVIEW

File: packages/core-dialog/core-dialog.js [hunk 1]
```diff
@@ -25,23 +25,37 @@ export default function dialog (dialogs, open) {
     dialog.setAttribute('aria-modal', true)
     dialog.setAttribute('aria-label', options.label || dialog.getAttribute('aria-label'))
     hasBackdrop || dialog.insertAdjacentElement('afterend', document.createElement('backdrop'))
+    options.strict && dialog.setAttribute('data-dialog-strict', options.strict)
```
Comment (eirikbacker) [inline packages/core-dialog/core-dialog.js]: you can set the attribute anyway, and I would suggest storing it in `UUID` instead:
```
dialog.setAttribute(UUID, Boolean(options.strict)) // Make sure it is true/false
```

File: packages/core-dialog/core-dialog.js [hunk 2]
```diff
@@ -25,23 +25,37 @@ export default function dialog (dialogs, open) {
...
     return dialog
   })
 }
 
+// Focus can move outside the dialog when in "strict"-mode, therefore
+// we need to ensure that if that happens we will refocus on the dialog.
+addEvent(UUID, 'focus', (event) => {
+  let hasDialog = false
+  for (let el = event.target; el; el = el.parentElement) {
+    if (el.hasAttribute(UUID)) return (hasDialog = true)
```
Comment (eirikbacker) [inline packages/core-dialog/core-dialog.js]: why do you need the `hasDialog` - as you do `return`, you exit the whole function all together anyway :)

File: packages/core-dialog/core-dialog.js [hunk 3]
```diff
@@ -25,23 +25,37 @@ export default function dialog (dialogs, open) {
...
+}, true)
+
 addEvent(UUID, 'click', (event) => {
   // This functions handles if the user clicked on a open or close button.
   // We need to do this loop in order for the close button to know which
   // dialog it should close.
   for (let el = event.target, isClose; el; el = el.parentElement) {
     const action = el.getAttribute(ATTR)
     const prev = el.previousElementSibling
+    const isStrict = prev && prev.hasAttribute(UUID) && Boolean(prev.getAttribute('data-dialog-strict'))
```
Comment (eirikbacker) [inline packages/core-dialog/core-dialog.js]: remove this

File: packages/core-dialog/core-dialog.js [hunk 4]
```diff
@@ -25,23 +25,37 @@ export default function dialog (dialogs, open) {
...
     isClose = isClose || action === 'close'
```
Comment (eirikbacker) [inline packages/core-dialog/core-dialog.js]: And change this to
```
change to:
```
isClose = isClose || (prev && prev.getAttribute(UUID) === 'false' && (el = prev)) || action === 'close'
```

File: packages/core-dialog/core-dialog.js [hunk 5]
```diff
@@ -25,23 +25,37 @@ export default function dialog (dialogs, open) {
...
 
     if (isClose) el.hasAttribute(UUID) && setOpen(el, false)
-    else if (prev && prev.hasAttribute(UUID)) setOpen(prev, false) // Click on backdrop
+    else if (!isStrict && prev && prev.hasAttribute(UUID)) setOpen(prev, false) // Click on backdrop
```
Comment (eirikbacker) [inline packages/core-dialog/core-dialog.js]: and then remove this

File: packages/core-dialog/core-dialog.js [hunk 6]
```diff
@@ -54,9 +68,10 @@ addEvent(UUID, 'transitionend', ({target}) => {
 addEvent(UUID, 'keydown', (event) => {
   const key = event.keyCode
   const dialog = !event.defaultPrevented && (key === KEYS.ESC || key === KEYS.TAB) && getTopLevelDialog()
+  const isStrict = dialog && Boolean(dialog.getAttribute('data-dialog-strict'))
```
Comment (eirikbacker) [inline packages/core-dialog/core-dialog.js]: change to:
```
const isClose = dialog && dialog.getAttribute(UUID) === 'false'
```

File: packages/core-dialog/core-dialog.js [hunk 7]
```diff
@@ -54,9 +68,10 @@ addEvent(UUID, 'transitionend', ({target}) => {
...
 
   if (dialog && key === KEYS.TAB) keepFocus(dialog, event)
-  if (dialog && key === KEYS.ESC) setOpen(dialog, false, event.preventDefault())
+  if (dialog && !isStrict && key === KEYS.ESC) setOpen(dialog, false, event.preventDefault())
```
Comment (eirikbacker) [inline packages/core-dialog/core-dialog.js]: ...and then change this to:
```
if (isClose && key === KEYS.ESC) setOpen(dialog, false, event.preventDefault())
```

File: packages/core-dialog/core-dialog.jsx [hunk 8]
```diff
@@ -6,14 +6,14 @@ const DEFAULTS = {hidden: null, onToggle: null}
...
+    dialog(this.el, exclude(this.props, DEFAULTS))
     this.el.addEventListener('dialog.toggle', this.props.onToggle)
   }
   componentWillUnmount () {
     this.el.removeEventListener('dialog.toggle', this.props.onToggle)
   }
-  componentWillReceiveProps ({open}) {
-    dialog(this.el, open)
+  componentWillReceiveProps (props) {
+    dialog(this.el, exclude(props, DEFAULTS))
```
Comment (eirikbacker) [inline packages/core-dialog/core-dialog.jsx]: why exclude DEFAULTS here?

File: packages/core-dialog/core-dialog.jsx [hunk 9]
```diff
@@ -6,14 +6,14 @@ const DEFAULTS = {hidden: null, onToggle: null}
 
 export default class Dialog extends React.Component {
   componentDidMount () {
-    dialog(this.el, this.props.open)
+    dialog(this.el, exclude(this.props, DEFAULTS))
```
Comment (eirikbacker) [inline packages/core-dialog/core-dialog.jsx]: why exclude DEFAULTS here?

File: packages/core-dialog/core-dialog.md [hunk 10]
```diff
@@ -135,7 +169,7 @@ coreDialog('.my-dialog', {open: true, label: 'A super dialog'})
 ```jsx
 import Dialog from '@nrk/core-dialog/jsx'
 
-<Dialog open={true|false} onToggle={function(){}} aria-label="Title of dialog">
+<Dialog open={true|false} strict={'true'| /*or omit attribute*/} onToggle={function(){}} aria-label="Title of dialog">
```
Comment (eirikbacker) [inline packages/core-dialog/core-dialog.md]: should be `strict={true|false}`

File: packages/core-dialog/core-dialog.md [hunk 11]
```diff
@@ -31,22 +39,35 @@ class DialogContainerTest extends React.Component {
     super(props)
     this.state = {
       open: false,
+      strictOpen: false,
       contentTitle: 'Dialog for JSX'
     }
     this.toggleDialog = this.toggleDialog.bind(this)
+    this.toggleStrictDialog = this.toggleStrictDialog.bind(this)
```
Comment (eirikbacker) [inline packages/core-dialog/core-dialog.md]: why do you need to `toggleStrictDialog`?

**Comment by wwalmnes** [comment]
Fixed issues with @eirikbacker

**Comment by eirikbacker** [comment]
Skal jeg merge denne @wwalmnes ? :)

**Pull request merged**